web/api/v1: Unit Tests fail nondeterministically on master

## Bug Report
**What did you do?**
* Checkout latest master branch (as of commit 0ae048ff ).
* Run `go test ./web/api/v1` multiple times, without test caching.

**What did you expect to see?**
The test should always pass.
**What did you see instead? Under which circumstances?**
Sometimes the test fails. It might need multiple runs to reproduce this.

```

--- FAIL: TestEndpoints (0.03s)
    --- FAIL: TestEndpoints/local (0.01s)
        api_test.go:1157: run 0	GET	"query=2&time=123.4"
        api_test.go:1157: run 0	POST	"query=2&time=123.4"
        api_test.go:1157: run 1	GET	"query=0.333&time=1970-01-01T00%3A02%3A03Z"
        api_test.go:1157: run 1	POST	"query=0.333&time=1970-01-01T00%3A02%3A03Z"
        api_test.go:1157: run 2	GET	"query=0.333&time=1970-01-01T01%3A02%3A03%2B01%3A00"
        api_test.go:1157: run 2	POST	"query=0.333&time=1970-01-01T01%3A02%3A03%2B01%3A00"
        api_test.go:1157: run 3	GET	"query=0.333"
        api_test.go:1157: run 3	POST	"query=0.333"
        api_test.go:1157: run 4	GET	"end=2&query=time%28%29&start=0&step=1"
        api_test.go:1157: run 4	POST	"end=2&query=time%28%29&start=0&step=1"
        api_test.go:1157: run 5	GET	"end=2&query=time%28%29&step=1"
        api_test.go:1157: run 5	POST	"end=2&query=time%28%29&step=1"
        api_test.go:1157: run 6	GET	"query=time%28%29&start=0&step=1"
        api_test.go:1157: run 6	POST	"query=time%28%29&start=0&step=1"
        api_test.go:1157: run 7	GET	"end=2&query=time%28%29&start=0"
        api_test.go:1157: run 7	POST	"end=2&query=time%28%29&start=0"
        api_test.go:1157: run 8	GET	"query=invalid%5D%5Bquery&time=1970-01-01T01%3A02%3A03%2B01%3A00"
        api_test.go:1157: run 8	POST	"query=invalid%5D%5Bquery&time=1970-01-01T01%3A02%3A03%2B01%3A00"
        api_test.go:1157: run 9	GET	"end=100&query=invalid%5D%5Bquery&start=0&step=1"
        api_test.go:1157: run 9	POST	"end=100&query=invalid%5D%5Bquery&start=0&step=1"
        api_test.go:1157: run 10	GET	"end=2&query=time%28%29&start=1&step=0"
        api_test.go:1157: run 10	POST	"end=2&query=time%28%29&start=1&step=0"
        api_test.go:1157: run 11	GET	"end=1&query=time%28%29&start=2&step=1"
        api_test.go:1157: run 11	POST	"end=1&query=time%28%29&start=2&step=1"
        api_test.go:1157: run 12	GET	"end=1489667272.372&query=time%28%29&start=148966367200.372&step=1"
        api_test.go:1157: run 12	POST	"end=1489667272.372&query=time%28%29&start=148966367200.372&step=1"
        api_test.go:1157: run 13	GET	"match%5B%5D=test_metric2"
        api_test.go:1157: run 13	POST	"match%5B%5D=test_metric2"
        api_test.go:1157: run 14	GET	"match%5B%5D=test_metric1%7Bfoo%3D~%22.%2Bo%22%7D"
        api_test.go:1157: run 14	POST	"match%5B%5D=test_metric1%7Bfoo%3D~%22.%2Bo%22%7D"
        api_test.go:1157: run 15	GET	"match%5B%5D=test_metric1%7Bfoo%3D~%22.%2Bo%24%22%7D&match%5B%5D=test_metric1%7Bfoo%3D~%22.%2Bo%22%7D"
        api_test.go:1157: run 15	POST	"match%5B%5D=test_metric1%7Bfoo%3D~%22.%2Bo%24%22%7D&match%5B%5D=test_metric1%7Bfoo%3D~%22.%2Bo%22%7D"
        api_test.go:1157: run 16	GET	"match%5B%5D=test_metric1%7Bfoo%3D~%22.%2Bo%22%7D&match%5B%5D=none"
        api_test.go:1157: run 16	POST	"match%5B%5D=test_metric1%7Bfoo%3D~%22.%2Bo%22%7D&match%5B%5D=none"
        api_test.go:1157: run 17	GET	"end=-1&match%5B%5D=test_metric2&start=-2"
        api_test.go:1157: run 17	POST	"end=-1&match%5B%5D=test_metric2&start=-2"
        api_test.go:1157: run 18	GET	"end=100001&match%5B%5D=test_metric2&start=100000"
        api_test.go:1157: run 18	POST	"end=100001&match%5B%5D=test_metric2&start=100000"
        api_test.go:1157: run 19	GET	"end=100000&match%5B%5D=test_metric2&start=-1"
        api_test.go:1157: run 19	POST	"end=100000&match%5B%5D=test_metric2&start=-1"
        api_test.go:1157: run 20	GET	"end=100&match%5B%5D=test_metric2&start=1"
        api_test.go:1157: run 20	POST	"end=100&match%5B%5D=test_metric2&start=1"
        api_test.go:1157: run 21	GET	"end=100000&match%5B%5D=test_metric2&start=1"
        api_test.go:1157: run 21	POST	"end=100000&match%5B%5D=test_metric2&start=1"
        api_test.go:1157: run 22	GET	"end=1&match%5B%5D=test_metric2&start=-1"
        api_test.go:1157: run 22	POST	"end=1&match%5B%5D=test_metric2&start=-1"
        api_test.go:1157: run 23	GET	""
        api_test.go:1157: run 23	POST	""
        api_test.go:1157: run 24	GET	""
        api_test.go:1157: run 25	GET	""
        api_test.go:1157: run 26	GET	"state=any"
        api_test.go:1157: run 27	GET	"state=active"
        api_test.go:1157: run 28	GET	"state=Dropped"
        api_test.go:1157: run 29	GET	"metric=go_threads"
        api_test.go:1157: run 30	GET	"match_target=%7Bjob%3D%22blackbox%22%7D"
        api_test.go:1157: run 31	GET	""
        api_test.go:1199: Response does not match, expected:
            [{"target":{"job":"test"},"metric":"go_threads","type":"gauge","help":"Number of OS threads created.","unit":""},{"target":{"job":"blackbox"},"metric":"prometheus_tsdb_storage_blocks_bytes","type":"gauge","help":"The number of bytes that are currently used for local storage by all blocks.","unit":""}]
            got:
            [{"target":{"job":"blackbox"},"metric":"prometheus_tsdb_storage_blocks_bytes","type":"gauge","help":"The number of bytes that are currently used for local storage by all blocks.","unit":""},{"target":{"job":"test"},"metric":"go_threads","type":"gauge","help":"Number of OS threads created.","unit":""}]
FAIL
FAIL	github.com/prometheus/prometheus/web/api/v1	0.062s
FAIL
```
**Environment**

* System information:
Linux 5.3.13-300.fc31.x86_64 x86_64

--- a/web/api/v1/api_test.go
+++ b/web/api/v1/api_test.go
@@ -540,7 +540,7 @@ func testEndpoints(t *testing.T, api *API, testLabelAPI bool) {
 		query    url.Values
 		response interface{}
 		errType  errorType
-		sorter   func(interface{}) interface{}
+		sorter   func(interface{})
 	}
 
 	var tests = []test{
@@ -1011,10 +1011,11 @@ func testEndpoints(t *testing.T, api *API, testLabelAPI bool) {
 					Unit:   "",
 				},
 			},
-			sorter: func(m interface{}) interface{} {
-				metrics := m.([]metricMetadata)
-				sort.Slice(metrics, func(i, j int) bool { return metrics[i].Metric < metrics[j].Metric })
-				return metrics
+			sorter: func(m interface{}) {
+				sort.Slice(m.([]metricMetadata), func(i, j int) bool {
+					s := m.([]metricMetadata)
+					return s[i].Metric < s[j].Metric
+				})
 			},
 		},
 		// Without a matching metric.
@@ -1170,12 +1171,11 @@ func testEndpoints(t *testing.T, api *API, testLabelAPI bool) {
 			res := test.endpoint(req.WithContext(ctx))
 			assertAPIError(t, res.err, test.errType)
 
-			data := res.data
 			if test.sorter != nil {
-				data = test.sorter(data)
+				test.sorter(res.data)
 			}
 
-			assertAPIResponse(t, data, test.response)
+			assertAPIResponse(t, res.data, test.response)
 		}
 	}
 }

--- a/web/api/v1/api_test.go
+++ b/web/api/v1/api_test.go
@@ -1198,6 +1198,8 @@ func assertAPIError(t *testing.T, got *apiError, exp errorType) {
 }
 
 func assertAPIResponse(t *testing.T, got interface{}, exp interface{}) {
+	t.Helper()
+
 	if !reflect.DeepEqual(exp, got) {
 		respJSON, err := json.Marshal(got)
 		if err != nil {

--- a/web/api/v1/api_test.go
+++ b/web/api/v1/api_test.go
@@ -27,6 +27,7 @@ import (
 	"net/url"
 	"os"
 	"reflect"
+	"sort"
 	"strings"
 	"testing"
 	"time"
@@ -539,6 +540,7 @@ func testEndpoints(t *testing.T, api *API, testLabelAPI bool) {
 		query    url.Values
 		response interface{}
 		errType  errorType
+		sorter   func(interface{}) interface{}
 	}
 
 	var tests = []test{
@@ -1009,6 +1011,11 @@ func testEndpoints(t *testing.T, api *API, testLabelAPI bool) {
 					Unit:   "",
 				},
 			},
+			sorter: func(m interface{}) interface{} {
+				metrics := m.([]metricMetadata)
+				sort.Slice(metrics, func(i, j int) bool { return metrics[i].Metric < metrics[j].Metric })
+				return metrics
+			},
 		},
 		// Without a matching metric.
 		{
@@ -1162,7 +1169,13 @@ func testEndpoints(t *testing.T, api *API, testLabelAPI bool) {
 			}
 			res := test.endpoint(req.WithContext(ctx))
 			assertAPIError(t, res.err, test.errType)
-			assertAPIResponse(t, res.data, test.response)
+
+			data := res.data
+			if test.sorter != nil {
+				data = test.sorter(data)
+			}
+
+			assertAPIResponse(t, data, test.response)
 		}
 	}
 }

